src/client/Microsoft.Identity.Client/Instance/Oidc/OidcRetrieverWithCache.cs

```diff
@@ -70,6 +75,58 @@ public static async Task<OidcMetadata> GetOidcAsync(
             }
         }
 
+        /// <summary>
+        /// Validates that the issuer in the OIDC metadata matches the authority.
+        /// </summary>
+        /// <param name="authority">The authority URL.</param>
+        /// <param name="issuer">The issuer from the OIDC metadata - the single source of truth.</param>
+        /// <exception cref="MsalServiceException">Thrown when issuer validation fails.</exception>
+        private static void ValidateIssuer(Uri authority, string issuer)
+        {
+
+            // Primary validation: check if authority begins with the issuer string (exact string comparison)
+            if (authority.AbsoluteUri.StartsWith(issuer, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            // Extract tenant for CIAM-like scenarios
```

Comment:
There are 2 ways to solve this: 

1. Assume that app developers will always use `WithAuthority` API for CIAM and AAD authorities. They would only use `WithOidcAuthority` for CIAM CUD. In which case, you only need block 88-90
2. OR, assume that app developers can mix and match. And that some will use `WithOidcAuthority(a)` with AAD and regular CIAM, in which case your logic needs more code for AAD.

If you go on the first path, then pls make sure to run Id.Web tests, to check that they do the right thing.

Reply:
Improved in the latest commit:
` // Extract tenant for CIAM scenarios. In a CIAM scenario the issuer is expected to have "{tenant}.ciamlogin.com" as the host, even when using a custom domain.`